feat(extensions): allow installing plugins based on access and env

debsmita1 opened the pull request:
## Hey, I just made a Pull Request!

## Resolves: 

https://issues.redhat.com/browse/RHIDP-6812
Part of: https://issues.redhat.com/browse/RHIDP-6811


- Disabled the installation for the production env. Added an alert for the production environment
- Integrated the fetch configuration API call
- Integrated the install API call
- Enable the install button for users with create access, for dev environment, when `extensions.installation.enabled` in the app-config is set to true, and when `extensions.installation.saveToSingleFile.file` in the app-config is configured correctly

If an Error occurs on installation
![Screenshot 2025-06-10 at 1 55 29 PM](https://github.com/user-attachments/assets/52761b9b-c04c-4ab6-869b-9fa9f8e6bda8)


For production env
![Screenshot 2025-06-09 at 1 54 31 PM](https://github.com/user-attachments/assets/291e5f25-cb3c-48eb-96aa-defec8c2d48c)



For development env, when installation is disabled or the extensions plugin configuration is missing in the app-config.yaml
![Screenshot 2025-06-09 at 3 42 07 PM](https://github.com/user-attachments/assets/30282f3c-ebe0-45ed-b024-4b97df735921)

![Screenshot 2025-06-06 at 12 20 22 AM](https://github.com/user-attachments/assets/ee3fc86f-c3f8-4714-b48b-e458ceb4c821)



For an incorrect config file name
![Screenshot 2025-06-10 at 1 52 00 PM](https://github.com/user-attachments/assets/ab459b55-47e5-4b02-8b9b-31b5c7e147ee)



For the missing config file
![Screenshot 2025-06-10 at 1 35 40 PM](https://github.com/user-attachments/assets/20420586-8e39-4c3e-becc-de12754ee67a)



For an invalid installation configuration, the configuration is fetched from the dynamic artifacts as a fallback option
![Screenshot 2025-06-10 at 1 54 29 PM](https://github.com/user-attachments/assets/0eda9c73-6749-4138-ad1b-5486d3b57734)

## How to test

Create a dynamic-plugins.yaml file
```
plugins:
  - package: ./dynamic-plugins/dist/red-hat-developer-hub-backstage-plugin-bulk-import-backend-dynamic
    disabled: false
  - package: ./dynamic-plugins/dist/red-hat-developer-hub-backstage-plugin-bulk-import
    disabled: true
    pluginConfig:
      dynamicPlugins:
        frontend:
          red-hat-developer-hub.backstage-plugin-bulk-import:
            appIcons:
              - name: bulkImportIcon3
                importName: BulkImportIcon3
            dynamicRoutes:
              - path: /bulk-import/repositories
                importName: BulkImportPage
                menuItem:
                  icon: bulkImportIcon
                  text: Bulk import
  - package: ./dynamic-plugins/dist/red-hat-developer-hub-backstage-plugin-marketplace
    disabled: false # DEBSMITA TEST
    pluginConfig:
      dynamicPlugins:
        frontend:
          red-hat-developer-hub.backstage-plugin-marketplace:
            appIcons:
              - name: marketplace
                importName: MarketplaceIcon1
            dynamicRoutes:
              - path: /extensions/catalog
                importName: DynamicMarketplacePluginRouter
            mountPoints:
              - mountPoint: internal.plugins/tab
                importName: DynamicMarketplacePluginContent
                config:
                  path: marketplace
                  title: Catalog
  - package: ./dynamic-plugins/dist/red-hat-developer-hub-backstage-plugin-marketplace-backend-dynamic
    disabled: false
  - package: ./dynamic-plugins/dist/backstage-community-plugin-topology
    disabled: true # UPDATED222
  - package: ./dynamic-plugins/dist/backstage-community-plugin-quay
    disabled: false
    # TODO: enable
    pluginConfig:
      dynamicPlugins:
        frontend:
          backstage-community.plugin-quay:
            mountPoints:
              - mountPoint: entity.page.image-registry/cards
                importName: QuayPage
                config:
                  layout:
                    gridColumn: 1 / -1
                  if:
                    anyOf:
                      - isQuayAvailable

```

app-config.yaml
```
extensions:
  ### Example for how to enable installation to a file.
  installation:
    enabled: true
    saveToSingleFile:
      file: <path-to>/dynamic-plugins.yaml
```

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/redhat-developer/rhdh-plugins/blob/main/CONTRIBUTING.md#creating-changesets))
- [x] Added or Updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

ShiranHi commented:
@debsmita1 thank you for implementing these changes! I have a few comments to help align the implementation more closely with the design:

1. Could we ensure the info-error bars consistently appear above the filters, as depicted in [the design](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s/Administration?node-id=9813-3994&t=70xrkjDiZAjB6bob-1)?
 2. The. error bar "Error occurs" should also be moved to the top of the page. Do we have any more descriptive informational text available beyond just "Error occurred"?
3. For the tooltip above the disabled "Install" button, could we reduce its width to prevent single words from wrapping to a new line?
4. Can we remove the prefix "Info:" that appears at the beginning of all info bars?
5. On the "Plugin installation is disabled in the production environment" info bar, could we also show users how to enable this setting, as specified in [the design](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s/Administration?node-id=9813-3994&t=70xrkjDiZAjB6bob-1)?
6. Please update the text for the "Configuration file doesn't exist" message to match [the design's](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s/Administration?node-id=10074-48618&t=70xrkjDiZAjB6bob-1) wording.
7. Please update the text for the "Invalid installation configuration" message to match [the design's](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s/Administration?node-id=10074-48841&t=70xrkjDiZAjB6bob-1) wording.
8. I think we missing the "Value in the configuration file is missing" error, right? If so, this is [the design](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s/Administration?node-id=10074-49069&t=70xrkjDiZAjB6bob-1).

ciiay commented on workspaces/marketplace/plugins/marketplace-backend/src/router.ts:
When test this PR on local RHDH, I got this error
```
backend:start: 2025-06-09T17:42:06.949Z extensions error Request failed with status 500 Cannot read properties of undefined (reading 'type') type="errorHandler" stack="TypeError: Cannot read properties of undefined (reading 'type')\n    at authorizeConditional (/Users/yicai/redhat/rhdh/dynamic-plugins-root/red-hat-developer-hub-backstage-plugin-marketplace-backend-dynamic/src/router.ts:99:20)\n    at async Promise.all (index 0)\n    at async <anonymous> (/Users/yicai/redhat/rhdh/dynamic-plugins-root/red-hat-developer-hub-backstage-plugin-marketplace-backend-dynamic/src/router.ts:251:47)" trace_id="f6589bb9b89b60794566dbdad9134d63" span_id="5893198a725e1f18" trace_flags="01"
```
which indicates that permission was undefined when calling this function
```
  const authorizeConditional = async (
    request: Request,
    permission:
      | ResourcePermission<'extensions-plugin' | 'extensions-package'>
      | BasicPermission,
  ) => {
    const credentials = await httpAuth.credentials(request);
    let decision: PolicyDecision;
    if (permission.type === 'resource') { // here
      decision = (
        await permissions.authorizeConditional([{ permission }], {
          credentials,
        })
      )[0];
    } else {
      decision = (
        await permissions.authorize([{ permission }], {
          credentials,
        })
      )[0];
    }

    return decision;
  };
```

So better we add an undefined check in this function.

Not sure if it's due to this error, all plugins are not installable to me like this
![image](https://github.com/user-attachments/assets/4589dedd-c6e9-4727-beaf-a8ad01a74d8d)

My `NODE_ENV` is set to `dev` locally, and still in the tooltip it says this is a production env.

debsmita1 replied:
I looked into the error caused by the `authorizeConditional` function, and it appears that when we test the changes in rhdh, we do not have the latest marketplace common package, which contains the permission information. As a result, `permission` is undefined.

ciiay replied:
I still think we better have a "undefined" check for the `permission` object, just in case. What do you think?

debsmita1 replied:
Sure, I will add that

ciiay reviewed (commented):
Tested in the dev app and it works as expected. When I tried to test it on RHDH, I got some backend error.

debsmita1 commented:
> > > 2. The. error bar "Error occurs" should also be moved to the top of the page. Do we have any more descriptive informational text available beyond just "Error occurred"?
> > 
> > 
> > This error is shown if the installation fails for any reason. I am displaying the message that the backend API returns incase of an error
> 
> Thanks, it is better this way, can we move it to the top too?

Updated the screenshots in the PR description

> 
> > > 4. Can we remove the prefix "Info:" that appears at the beginning of all info bars?
> > 
> > 
> > The Backstage’s Warning Panel that renders this component adds the “severity” in the beginning of the title
> 
> Can we get rid of it? It is clear this is an info based on the background color
> 

I am afraid we can’t with the Backstage’s warning panel component that renders this kind of accordion-style error panel, I will have to create a custom component to get rid of this


> > > 5. On the "Plugin installation is disabled in the production environment" info bar, could we also show users how to enable this setting, as specified in [the design](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s/Administration?node-id=9813-3994&t=70xrkjDiZAjB6bob-1)?
> > 
> > 
> > The feature will remain disabled for production in this release (https://issues.redhat.com/browse/RHIDP-6240) , but in the dev mode, the user can enable/disable installation with the app-config configuration
> 
> Thanks, it will be great to add this information to the info bar, with showing how the can enable it with a code block, what do you think?
> 

I have added it already, for production env I am showing “The plugin installation is disabled in the production environment” and in development mode, if the extensions installation is disabled, I am showing “Plugin installation is disabled” with hightlighted code block on how to enable it. I have screenshots attached in the PR description

> > > I think we missing the "Value in the configuration file is missing" error, right? If so, this is [the design](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s/Administration?node-id=10074-49069&t=70xrkjDiZAjB6bob-1).
> 
> Thanks! Can we please align the text to be like in the Figma file?
> 
> Also, did we change these? 6. Please update the text for the "Configuration file doesn't exist" message to match [the design's](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s/Administration?node-id=10074-48618&t=70xrkjDiZAjB6bob-1) wording. 7. Please update the text for the "Invalid installation configuration" message to match [the design's](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s/Administration?node-id=10074-48841&t=70xrkjDiZAjB6bob-1) wording.

Done PTAL

ShiranHi commented:
> * If we're not planning to show any explanation, should we hide the button entirely instead? Right now, it’s a bit confusing to see a disabled button with no context.

From what I remember there is a tooltip implemented ([this one](https://www.figma.com/design/E6C9X3OSVzlWkW7ZsbxN7s/Administration?node-id=9813-4440&t=w3vqgEHd4GZpHesy-1)).

>In this case, "Back" or "Close" might be a more intuitive label. That said, if we want to keep consistent buttons across states, I totally understand.

Yeah, good point, we can change it to be "Back".
